### DISCUSSION TITLE
What is the correct way to exit the app?

### DISCUSSION THREAD

**Discussion by moxaj** [tauri-apps/tauri#3555] [Q&A] [answered]
If I simply use `AppHandle::exit`, I get

```
[0224/220617.016:ERROR:window_impl.cc(114)] Failed to unregister class Chrome_WidgetWin_0. Error = 0
```

FWIW, I have 2 separate windows, 1 not visible, and a system tray as well. And, if I do this while development (`npm run tauri dev`), I also get a huge powershell exception :(

**Comment by FabianLars** [comment, marked as answer]
AppHandle::exit is correct. Of course not seeing errors would be nicer, but these ones are nothing to worry about 🤷

**Reply by moxaj** [reply]
Sorry to resurrect this, but I'm still having issues with closing the application. Even in a release build, invoking `AppHandle::exit` closes all the windows, kills the system tray icon, but after ~10 seconds another window appears: it has a taskbar icon (the generic windows one), the window itself is a small black rectangle and is unresponsive.

**Reply by moxaj** [reply]
Ok, turns out you shouldn't call `AppHandle::exit` on the main thread. ¯\_(ツ)_/¯

**Comment by milotheirself** [comment]
If the [default behaviour](https://github.com/tauri-apps/tauri/issues/2287#issuecomment-886108338) isn't prevented, an probably neater alternative to using `AppHandle::exit` is to close any open windows and let the Taurus close itself. This will not throw the mentioned error.
